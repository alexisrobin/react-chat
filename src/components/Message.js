import React from 'react';
import './Message.scss';
import { Container, Row, Col } from 'reactstrap';

const Message = ({ message, senderName }) => {

    const convertISOStringToReadableDateString = ISOString =>
        new Intl.DateTimeFormat('en-US', { hour: '2-digit', minute: '2-digit', year: 'numeric', month: 'short', day: '2-digit' }).format(new Date(Date.parse(ISOString)));

    return (
        <Container>
            <Row className="message-content rounded p-2">
                {message.content}
            </Row>
            <Row className="message-timestamp">
                <Col className="text-right">
                    sent by {senderName} - {convertISOStringToReadableDateString(message.timestamp)}
                </Col>
            </Row>
        </Container>
    );
}

export default Message;