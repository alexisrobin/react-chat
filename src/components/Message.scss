.message-content {
    background-color: #ddd;
}

.message-timestamp {
    color: #aaa;
    font-size: 0.9rem;
    font-style: italic;
}