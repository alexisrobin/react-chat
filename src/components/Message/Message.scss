@import '../../const.scss';

.message-sent {
    .message-content {
        background-color: $primary;
        color: white;
    }
    .message-timestamp {
        text-align: right;
    }
}

.message-received {
    .message-content {
        background-color: #ddd;
        color: black;
    }
    .message-timestamp {
        text-align: left;
    }
}

.message-timestamp {
    color: #aaa;
    font-size: 0.9rem;
    font-style: italic;
}

.message-unreadable {
    text-align: center;
    font-style: italic;
}