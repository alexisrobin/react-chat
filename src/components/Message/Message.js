import React from 'react';
import './Message.scss';
import { Container, Row, Col } from 'reactstrap';
import { FadeTransform, } from 'react-animation-components';


const convertDateFromISOStringToReadableString = ISOString =>
    new Intl.DateTimeFormat('en-US', { hour: '2-digit', minute: '2-digit', year: 'numeric', month: 'short', day: '2-digit' }).format(new Date(Date.parse(ISOString)));

/**
 * Basic chat's message adapting it's view depending of its nature (sent, received).
 */
const Message = ({ message, sender, currentUserId }) => {
    if (!message) {
        return (
            <div className="message-unreadable">
                Unreadable message
            </div>
        );
    }
    else
        return (
            <FadeTransform
                in
                transformProps={
                    sender.id === currentUserId ?
                        {
                            enterTransform: 'translateX(0px)',
                            exitTransform: 'translateX(-100px)'
                        } :
                        {
                            enterTransform: 'translateX(0px)',
                            exitTransform: 'translateX(100px)'
                        }
                }>
                <Container className={sender.id === currentUserId ? "message-sent" : "message-received"}>
                    <Row className="message-content rounded p-2">
                        {message.content}
                    </Row>
                    <Row className="message-timestamp">
                        <Col>
                            sent by {sender.name} - {convertDateFromISOStringToReadableString(message.timestamp)}
                        </Col>
                    </Row>
                </Container>
            </FadeTransform>
        );
}

export default Message;
export { convertDateFromISOStringToReadableString };