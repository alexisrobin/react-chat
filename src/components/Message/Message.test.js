import React from 'react';
import { expect } from 'chai'
import { shallow } from 'enzyme'
import Message, { convertDateFromISOStringToReadableString } from './Message';

const message = {
    content: "testing message",
    timestamp: "2016-10-30T10:28:05"
};

describe('<Message />', () => {
    it('should match its reference snapshot', () => {
        const wrapper = shallow(
            <Message
                message={message}
                senderName={"Max"} />
        );

        expect(wrapper).to.matchSnapshot()
      })

    it('should renders properly message and timestamp', () => {
        const wrapper = shallow(
            <Message
                message={message}
                senderName={"Max"} />
        );

        expect(wrapper.contains(message.content)).to.equal(true);
        expect(wrapper.contains(convertDateFromISOStringToReadableString(message.timestamp))).to.equal(true);
    })

    it('should renders a message unreadable block', () => {
        const wrapper = shallow(
            <Message />
        );

        expect(wrapper.find('.message-unreadable')).to.have.length(1);
    })
})