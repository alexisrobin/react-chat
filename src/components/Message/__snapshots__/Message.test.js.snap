// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> should match its reference snapshot 1`] = `
<Container
  tag="div"
>
  <Row
    className="message-content rounded p-2"
    style={
      Object {
        "backgroundColor": undefined,
        "color": undefined,
      }
    }
    tag="div"
  >
    testing message
  </Row>
  <Row
    className="message-timestamp"
    tag="div"
  >
    <Col
      className="text-right"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      sent by 
      Max
       - 
      Oct 30, 2016, 10:28 AM
    </Col>
  </Row>
</Container>
`;
